Trilinos nightly compilation failures with Kokkos' develop branch

**Describe the bug**

One of the recent merges to Kokkos is breaking compilation of Tpetra when built against Kokkos' develop branch

**Error message**:
```
15:51:24 /home/jenkins/blake-new/workspace/KokkosEco_Trilinos_Blake_Clang1001_Serial_NoMPI_debug-develop/Trilinos/packages/tpetra/core/src/Tpetra_MultiVector_def.hpp:528:9: error: no matching function for call to 'checkGlobalDualViewValidity'
15:51:24         checkGlobalDualViewValidity (&gblErrStrm, view, verbose,
15:51:24         ^~~~~~~~~~~~~~~~~~~~~~~~~~~
15:51:24 /home/jenkins/blake-new/workspace/KokkosEco_Trilinos_Blake_Clang1001_Serial_NoMPI_debug-develop/Build/packages/tpetra/core/src/Tpetra_MultiVector_LONG_LONG_INT_LONG_LONG_SERIAL.cpp:74:3: note: in instantiation of member function 'Tpetra::MultiVector<long long, int, long long, Tpetra::KokkosCompat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> >::MultiVector' requested here
15:51:24   TPETRA_MULTIVECTOR_INSTANT( longlong, int, longlong, Tpetra_KokkosCompat_KokkosSerialWrapperNode )
15:51:24   ^
15:51:24 /home/jenkins/blake-new/workspace/KokkosEco_Trilinos_Blake_Clang1001_Serial_NoMPI_debug-develop/Trilinos/packages/tpetra/core/src/Tpetra_MultiVector_def.hpp:4929:18: note: expanded from macro 'TPETRA_MULTIVECTOR_INSTANT'
15:51:24   template class MultiVector< SCALAR , LO , GO , NODE >; \
15:51:24                  ^
15:51:24 /home/jenkins/blake-new/workspace/KokkosEco_Trilinos_Blake_Clang1001_Serial_NoMPI_debug-develop/Trilinos/packages/tpetra/core/src/Tpetra_Details_checkView.hpp:155:1: note: candidate template ignored: failed template argument deduction
15:51:24 checkGlobalDualViewValidity
...
```

These commits merges since the previous passing build:
dual view: update template types (#6085) ([details](https://jenkins-son.sandia.gov/job/KokkosEco_Trilinos_Blake_Clang1001_Serial_NoMPI_debug-develop/119/changes#1767bfe3e584377f1470e08b40f62ff7f5cf85cc))
    sorting an empty view should exit early and not fail (#6130) ([details](https://jenkins-son.sandia.gov/job/KokkosEco_Trilinos_Blake_Clang1001_Serial_NoMPI_debug-develop/119/changes#b86d73a2b84c3856043539842d5c0f1169663c47))



**Please also include the following items to support reproducing the bug**
**Reproducer** (Blake testbed, SKX arch, clang/10 Serial build):
```bash
# Repos
git clone -b develop https://github.com/trilinos/Trilinos.git
git clone -b develop https://github.com/kokkos/kokkos.git
git clone -b develop https://github.com/kokkos/kokkos-kernels.git
# Symbolic link to your kokkos and kokkos-kernels repos in Trilinos source directory for source override
cd Trilinos
ln -s <path-to-your-repo>/kokkos kokkos
ln -s <path-to-your-repo>/kokkos-kernels kokkos-kernels

cd $HOME
mkdir -p build
cd build

module load cmake/3.25.2 clang/10.0.1 openblas/0.3.13/gcc/10.2.0

cmake \
 -D CMAKE_INSTALL_PREFIX="${TRILINOS_INSTALL_DIR}" \
 -D CMAKE_CXX_STANDARD="17" \
 -D CMAKE_CXX_FLAGS="-g" \
 -D CMAKE_BUILD_TYPE=DEBUG \
 -D BUILD_SHARED_LIBS:BOOL=ON \
\
 -D TPL_ENABLE_MPI=OFF \
\
 -D TPL_ENABLE_BLAS:STRING=ON \
  -D BLAS_LIBRARY_DIRS:FILEPATH=${BLAS_ROOT}/lib \
  -D BLAS_LIBRARY_NAMES:STRING="openblas" \
 -D TPL_ENABLE_LAPACK:STRING=ON \
  -D LAPACK_INCLUDE_DIRS:FILEPATH="${LAPACK_ROOT}/include" \
  -D LAPACK_LIBRARY_DIRS:FILEPATH=${LAPACK_ROOT}/lib \
  -D LAPACK_LIBRARY_NAMES:STRING="openblas" \
\
 -D Trilinos_ENABLE_TESTS=ON \
 -D Trilinos_ENABLE_EXAMPLES=ON \
 -D Trilinos_ENABLE_COMPLEX=ON \
\
 -D Trilinos_ENABLE_Ifpack2=ON \
  -D Ifpack2_ENABLE_TESTS=ON \
 -D Trilinos_ENABLE_Amesos2=ON \
  -D Amesos2_ENABLE_TESTS=ON \
 -D Trilinos_ENABLE_Kokkos=ON \
 -D Kokkos_ENABLE_SERIAL=ON \
 -D Kokkos_ARCH_SKX=ON \
  -D Kokkos_ENABLE_TESTS=ON \
 -D Trilinos_ENABLE_KokkosKernels=ON \
  -D KokkosKernels_ENABLE_TESTS=ON \
 -D Trilinos_ENABLE_Tpetra=ON \
  -D Tpetra_ENABLE_TESTS=ON \
 -D Trilinos_ENABLE_Sacado=ON \
  -D Sacado_ENABLE_TESTS=ON \
 -D Trilinos_ENABLE_Stokhos=ON \
  -D Stokhos_ENABLE_TESTS=ON \
 -D Trilinos_ENABLE_Zoltan2=ON \
  -D Zoltan2_ENABLE_TESTS=ON \
 -D Trilinos_ENABLE_Intrepid2=ON \
  -D Intrepid2_ENABLE_TESTS=OFF \
 -D Trilinos_ENABLE_Belos=ON \
  -D Belos_ENABLE_TESTS=ON \
 -D Trilinos_ENABLE_Anasazi=ON \
  -D Anasazi_ENABLE_TESTS=ON \
 -D Trilinos_ENABLE_Teuchos=ON \
  -D Teuchos_ENABLE_TESTS=ON \
 -D Trilinos_ENABLE_MueLu=ON \
  -D MueLu_ENABLE_TESTS=ON \
 -D Trilinos_ENABLE_Panzer=ON \
  -D Panzer_ENABLE_TESTS=ON \
 -D Trilinos_ENABLE_Phalanx=ON \
  -D Phalanx_ENABLE_TESTS=ON \
 -D Kokkos_SOURCE_DIR_OVERRIDE:STRING=kokkos \
 -D KokkosKernels_SOURCE_DIR_OVERRIDE:STRING=kokkos-kernels \
$TRILINOS_DIR

```

--- a/containers/src/Kokkos_DualView.hpp
+++ b/containers/src/Kokkos_DualView.hpp
@@ -87,8 +87,14 @@ inline const Kokkos::Cuda& get_cuda_space(const NonCudaExecSpace&) {
 
 }  // namespace Impl
 
+#ifdef KOKKOS_ENABLE_DEPRECATED_CODE_4
+template <class DataType, class Arg1Type = void, class Arg2Type = void,
+          class Arg3Type = void>
+class DualView;
+#else
 template <class DataType, class... Properties>
 class DualView;
+#endif
 
 template <class>
 struct is_dual_view : public std::false_type {};
@@ -102,21 +108,35 @@ struct is_dual_view<const DualView<DT, DP...>> : public std::true_type {};
 template <class T>
 inline constexpr bool is_dual_view_v = is_dual_view<T>::value;
 
+#ifdef KOKKOS_ENABLE_DEPRECATED_CODE_4
+template <class DataType, class Arg1Type, class Arg2Type, class Arg3Type>
+class DualView : public ViewTraits<DataType, Arg1Type, Arg2Type, Arg3Type> {
+  template <class, class, class, class>
+#else
 template <class DataType, class... Properties>
 class DualView : public ViewTraits<DataType, Properties...> {
   template <class, class...>
+#endif
   friend class DualView;
 
  public:
   //! \name Typedefs for device types and various Kokkos::View specializations.
   //@{
-  using traits = ViewTraits<DataType, Properties...>;
+#ifdef KOKKOS_ENABLE_DEPRECATED_CODE_4
+  using traits = ViewTraits<DataType, Arg1Type, Arg2Type, Arg3Type>;
+#else
+  using traits      = ViewTraits<DataType, Properties...>;
+#endif
 
   //! The Kokkos Host Device type;
   using host_mirror_space = typename traits::host_mirror_space;
 
   //! The type of a Kokkos::View on the device.
-  using t_dev = View<typename traits::data_type, Properties...>;
+#ifdef KOKKOS_ENABLE_DEPRECATED_CODE_4
+  using t_dev = View<typename traits::data_type, Arg1Type, Arg2Type, Arg3Type>;
+#else
+  using t_dev       = View<typename traits::data_type, Properties...>;
+#endif
 
   /// \typedef t_host
   /// \brief The type of a Kokkos::View host mirror of \c t_dev.
@@ -124,7 +144,12 @@ class DualView : public ViewTraits<DataType, Properties...> {
 
   //! The type of a const View on the device.
   //! The type of a Kokkos::View on the device.
+#ifdef KOKKOS_ENABLE_DEPRECATED_CODE_4
+  using t_dev_const =
+      View<typename traits::const_data_type, Arg1Type, Arg2Type, Arg3Type>;
+#else
   using t_dev_const = View<typename traits::const_data_type, Properties...>;
+#endif
 
   /// \typedef t_host_const
   /// \brief The type of a const View host mirror of \c t_dev_const.

--- a/containers/unit_tests/CMakeLists.txt
+++ b/containers/unit_tests/CMakeLists.txt
@@ -49,6 +49,7 @@ endforeach()
 
 SET(COMPILE_ONLY_SOURCES
   TestCreateMirror.cpp
+  TestDualViewParameterPack.cpp
 )
 KOKKOS_ADD_EXECUTABLE(
   ContainersTestCompileOnly

--- a/containers/unit_tests/TestDualViewParameterPack.cpp
+++ b/containers/unit_tests/TestDualViewParameterPack.cpp
@@ -1,0 +1,43 @@
+//@HEADER
+// ************************************************************************
+//
+//                        Kokkos v. 4.0
+//       Copyright (2022) National Technology & Engineering
+//               Solutions of Sandia, LLC (NTESS).
+//
+// Under the terms of Contract DE-NA0003525 with NTESS,
+// the U.S. Government retains certain rights in this software.
+//
+// Part of Kokkos, under the Apache License v2.0 with LLVM Exceptions.
+// See https://kokkos.org/LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//@HEADER
+
+#include <Kokkos_Core.hpp>
+#include <Kokkos_DualView.hpp>
+
+namespace {
+
+template <class DataType, class Arg1Type = void, class Arg2Type = void,
+          class Arg3Type = void>
+void not_supported_anymore(
+    Kokkos::DualView<DataType, Arg1Type, Arg2Type, Arg2Type> x) {
+  static_assert(Kokkos::is_dual_view_v<decltype(x)>);
+}
+
+template <class DataType, class... Properties>
+void prefer_instead(Kokkos::DualView<DataType, Properties...> x) {
+  static_assert(Kokkos::is_dual_view_v<decltype(x)>);
+}
+
+using KDV = Kokkos::DualView<int*>;
+
+#ifdef KOKKOS_ENABLE_DEPRECATED_CODE_4
+static_assert(
+    std::is_void_v<decltype(not_supported_anymore(std::declval<KDV>()))>);
+#endif
+
+static_assert(std::is_void_v<decltype(prefer_instead(std::declval<KDV>()))>);
+
+}  // namespace

